msal/src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -124,6 +126,10 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
             var tenantId = Authority.CreateAuthority(requestParams.TenantUpdatedCanonicalAuthority, false)
                 .GetTenantId();
 
+            IdToken idToken = IdToken.Parse(response.IdToken);
+
+            var preferredUsername = !String.IsNullOrWhiteSpace(idToken?.PreferredUsername)? idToken.PreferredUsername : notReturnedFromServer;
```

Comment:
I was not explaining my self clearly enough. With the change I proposed it will still work with the update from B2C, however what you have removed now is the need to have preferred username specified for all other scenarios as well. Did you validate that this is all good? Are you sure we will not run into issues (at least usability issues) ?

Reply:
Yes, these changes were made with the guidance of @SomkaPe. The removal of the prefered_userName requirement will allow the cache to store tokens with this field missing. However, the token lookup operations do not depend on the preferred username so there will be no change in the behavior of caching operations. I ran this new change through our caching scenarios and no issues were found. We can follow up tomorrow during stand up.